added nextauth migration guide (#665)
clerk/clerk-docs: docs/guides/authjs-migration.mdx

---

> title: Migrating to Clerk from Auth.js
> description: Migrating from Auth.js

Reviewer: ```suggestion
title: Migrate from Auth.js to Clerk
description: Learn how to migrate an application using Auth.js to use Clerk for authentication.
```

---

<TutorialHero
  framework="authjs"
  exampleRepoTitle="Migration Script Repository"
  exampleRepo={[ 
    {
      title: "Migration Script",
      link: "https://github.com/clerk/migration-script"
      
    }
  ]}

  beforeYouStart={[
    {
      title: "Set up a Clerk application",
      link: "https://clerk.com/docs/quickstarts/setup-clerk",
    }
  ]}
>

- Install `@clerk/nextjs`
- Set environment keys
- Wrap your Next.js app in `<ClerkProvider />`
- Setup Sign-up and Sign-in UI
- Protect your app
- Read user and session data
- User IDs as Foreign Keys
- Create a production instance
- Migrate users to a production instance

> - Further support

Reviewer: ```suggestion
- Find further support
```

</TutorialHero>

> This guide shows how to migrate an application using Auth.js (formerly NextAuth.js) to use Clerk for authentication.

Reviewer: We need an H2 here, it can either be something like "Introduction" or "Overview" :)
I use "Introduction" on other pages but what ideas do you have for a heading here?

```suggestion
## Introduction

This guide shows how to migrate an application using Auth.js (formerly NextAuth.js) to use Clerk for authentication.
```

#### Account Portal

Account Portal is the fastest way to authenticate your app. Clerk's Account Portal hosts the `<SignIn />`, `<SignUp />`, `<UserProfile />`, and other components for your application. Read more about [Account Portal](/docs/account-portal/getting-started). 

To use the Account Portal, remove the routes that mount the Auth.js sign-in and sign-up UI. Replace the links to those routes with the [`<SignInButton>`](/docs/components/unstyled/sign-in-button) or [`<SignUpButton>`](/docs/components/unstyled/sign-out-button) components.

#### Self-hosted UI

If you want sign-in and sign-up pages in your own application, you can use [Clerk's pre-built components](/docs/references/nextjs/custom-signup-signin-pages), or implement the UI yourself using [Clerk's hooks](/docs/custom-flows/overview).

#### Custom UI

For fully custom UI, replace the existing functionality with [appropriate custom flows](/docs/custom-flows/overview#custom-flows). With Custom flows, you leverage Clerk's React hooks `useSignIn` and `useSignUp` to build fully custom UI that handles authentication with email and password, OAuth, magic links, OTP or any other authentication option Clerk offers.

> ### Protect your app

Reviewer: ```suggestion
### Protect your app

With Clerk, you can control access to your application in a few different ways. One way is to use Clerk's Middleware to protect your entire application, or specific routes. Another way is to use Clerk's components to conditionally render UI based on the user's authentication state. You can hide or show UI based on whether the user is signed in or not.
```

> #### Middleware

Reviewer: ```suggestion
#### Control access to your app with Clerk Middleware
```

You will need to remove the Auth.js Middleware from your application, and replace it with the Clerk Middleware helper. Auth's middleware always rejects unauthorized requests. You may have additionally configured the Next.js middleware config to protect specific private routes. You will need to make note of this configuration so you can recreate it.

1. Remove the Auth.js Middleware

```tsx filename="middleware.ts"
export { default } from "next-auth/middleware"

export const config = { matcher: [/** private routes */] }
```
2. Add Clerk's `authMiddleware()`

> The example below is the starting Middleware configuration. Please read the [`authMiddleware()`](/docs/references/nextjs/auth-middleware) documentation to learn how to configuration your Middleware. Clerk's Middleware will, by default, run for your entire application and gives you fine-grained control over handling the authenticated state. 

Reviewer: ```suggestion
Clerk's Middleware gives you fine-grained control over handling the authenticated state and will, by default, run for your entire application. 

The example below is a basic configuration that protects all routes. You will need to edit this to allow specific routes to be public as needed. Read the [`authMiddleware()`](/docs/references/nextjs/auth-middleware) documentation to learn more about how you can configure your Middleware.
```

```tsx filename="middleware.ts"
import { authMiddleware } from "@clerk/nextjs";
 
// This example protects all routes including api/trpc routes
// Please edit this to allow other routes to be public as needed.
// See https://clerk.com/docs/references/nextjs/auth-middleware for more information about configuring your Middleware
export default authMiddleware({});
 
export const config = {
  matcher: ["/((?!.+\\.[\\w]+$|_next).*)", "/", "/(api|trpc)(.*)"],
};
```

> 3. Use Clerk's control components to control access

Reviewer 2: The `3.` should remain, or the previous numbered items should change to `####`. This is meant to be part of the same list.
```
1. Remove the Auth.js Middleware
2. Add Clerk's `authMiddleware()`
```

To conditionally render UI when the user is signed in, wrap it with [`<SignedIn>`](/docs/components/control/signed-in).

To conditionally render UI when the user is NOT signed in, wrap it with [`<SignedOut>`](/docs/components/control/signed-out).

```tsx filename="app/page.tsx"
import { SignedIn, SignedOut } from "@clerk/nextjs";
 
export default function Home() {
  return (
    <div>
      <SignedOut>
        <p>This content is public. Only signed out users can see this.</p>
      </SignedOut>
      <SignedIn>
        <p>This content is private. Only signed in users can see this.</p>
      </SignedIn>
    </div>
  )
}
```

### Read user and session data

#### Server side

Replace any `getServerSession(req, res, authOptions)` with Clerk's helpers.

```tsx filename="app/page.tsx"
import { getServerSession } from "next-auth";
import { authOptions } from "@/pages/api/auth/[...nextauth]";

> export default function Home() {

Reviewer: ```suggestion
export default function Page() {
```

  const data = await getServerSession(authOptions);
  console.log(data)

  return (
    <p>Home Page</p>
  )
}
```

Use [`auth()`](/docs/references/nextjs/auth#auth) if you are using App Router or [`getAuth()`](/docs/references/nextjs/get-auth#get-auth) if you are using Pages Router to read your user data.

<CodeBlockTabs type="installer" options={["App Router", "Pages Router"]}>

```tsx failename="app/page.tsx"
import { auth, currentUser } from "@clerk/nextjs";
 
export default async function Page() {
 
  const { userId } = auth();
  console.log(userId)

  return (
    <p>Home Page</p>
  )

}
```

> ```tsx filename="app/page.tsx"

Reviewer: I think is supposed to be a different filename for pages router

export async function getServerSideProps(context) {
  const session = getAuth(context.req);
	
	return { props: { ...buildClerkProps(ctx.req) } };
};
```
</CodeBlockTabs>

#### Client Side

Replace Auth's `useSession` with Clerk's hooks.

```tsx filename="app/page.tsx"
'use client'
import { useSession } from "next-auth/react"

export default function Home() {
  const { data: session } = useSession()
  console.log(session)

  return (
    <p>Home Page</p>
  )
}
```

> Use [`useAuth()`](/docs/references/react/use-auth) if you need basic authentication information or [`useUser()`](/docs/references/react/use-user) if you need the full user object.

Reviewer: ```suggestion
Clerk's [`useAuth()`](/docs/references/react/use-auth) hook can be used to retrieve basic authentication information. Clerk's [`useUser()`](/docs/references/react/use-user) hook can be used to retrieve the full [`User`](https://clerk.com/docs/references/javascript/user/user#user) object, which includes information about the user, such as their first name, emails, phone numbers, and more.

Let's see what your code might look like when you replace Auth.js's `useSession()` hook with Clerk's hooks:
```

```tsx filename="app/page.tsx"
'use client'
import { useAuth, useUser } from "@clerk/nextjs";

export default function Home() {
  const { userId, sessionId } = useAuth();
  const { isSignedIn, user } = useUser();
  console.log(userId, sessionId, isSignedIn, user)

  return (
    <p>Home Page</p>
  )
}
```

### User IDs as Foreign Keys  

> When you migrate to Clerk, you will likely need to resolve the foreign key that you used in your database. If you used the `userId` from NextAuth.js, you could resolve this issue with one of the two following options.

Reviewer: ```suggestion
When you migrate to Clerk, you will likely need to resolve the foreign key that you used in your database. If you used the `userId` from NextAuth.js, you could resolve this issue with one of the following two options:

- [Use Clerk's `externalId` field](#use-clerk-s-external-id-field)
- Update your Auth.js database(#update-your-auth-js-database)
```

#### Use Clerk's `externalId` field

> When you migrate user data from Auth's database to Clerk, Clerk generates new user IDs for each user. If you are using existing user IDs as foreign keys in your database (e.g. in a `user_id` column), you can save those IDs as the user's `externalId` in Clerk. This `externalId` can be included in the session token by adding the following [customization](/docs/backend-requests/making/custom-session-token). The following example will return the `externalId` if one is present, and if not will return the Clerk `userId`.

Reviewer: ```suggestion
When you migrate user data from Auth.js's database to Clerk, Clerk generates new user IDs for each user. If you are using existing user IDs as foreign keys in your database (e.g. in a `user_id` column), you can save those IDs as the user's `externalId` in Clerk. This `externalId` can be included in the session token by adding the following [customization](/docs/backend-requests/making/custom-session-token). The following example will set the user's ID to be `externalId` if one is present, otherwise, it will use the user's original ID.
```

Reviewer 2: I modified this slightly:
- changed 'Auth.js's database to Clerk' to 'Auth.js to Clerk' to remove the idea that the DB is just Auth.js's and not the app's.

```json
{
	"userId": "{{user.external_id || user.id}}"
}
``` 

In your application, you can access this combined `userId` via session claims.

> ```tsx filename="app/page.tsx"
> import { auth } from "@clerk/nextjs";
>  
> export default async function Page() {
>  
>   const { sessionClaims: { userId } } = auth()
>   console.log(userId)
> 
>   return (
>     <p>Home Page</p>
>   )
> 
> }
> ```

Reviewer: Not sure if this is totally right, but the idea here is to have examples for both app and pages router (to stay consistent with the rest of this guide)

````suggestion
<Tabs type="router" items={["App Router", "Pages Router"]}>
<Tab>
To access the `userId` from the session claims, you can use the `auth()` helper.
```tsx filename="app/page.tsx"
import { auth } from "@clerk/nextjs";
 
export default async function Page() {
 
  const { sessionClaims: { userId } } = auth()
  console.log(userId)

  return (
    <p>Home Page</p>
  )

}
```
</Tab>

<Tab>
To access the `userId` from the session claims, you can use the `getAuth()` helper.

```tsx filename="pages/index.tsx"
import { getAuth } from "@clerk/nextjs/server";
import { NextRequest } from "next/server";
 
export default async function Home(req: NextRequest) {
 
  const { sessionClaims: { userId } } = getAuth(req)
  console.log(userId)

  return (
    <p>Home Page</p>
  )

}
```
</Tab>
</Tabs>
````

Reviewer 2: Added the tabs, but swapped out the Pages Router example for something appropriate.

<Callout type="info">

> You can not access the above from the client side. If you are using one of Clerk's hooks then you will need to check if `externalID` exists and if not then read the `userId`. 

Reviewer: ```suggestion
You can not access the above from the client side. If you are using one of Clerk's hooks, then you will need to check if `externalID` exists. If it doesn't, then read the `userId`. 
```

</Callout>

> #### Update your database

Reviewer: ```suggestion
#### Update your Auth.js database 
```

Reviewer 2: The database isn't really Auth.js's or Clerk's, but the application. Calling it `Auth.js database` could be confusing and isn't accurate.

Alternatively, after the data migration, you can update all the user IDs stored in your database as a foreign key to the new Clerk user IDs.

You can read more about user IDs and user data migration in the [Migration Script README](https://github.com/clerk/migration-script?tab=readme-ov-file#handling-the-foreign-key-constraint).

> ### Create a production instance

Reviewer: ```suggestion
### Create a Clerk production instance
```

> Every Clerk application has a Development and a Production instance. Before you start migrating user data, you will configure production instance and migrate your users directly into that instance. The [Deploying to Production](/docs/deployments/overview#deploying-to-production) page covers creating a production instance.
> 
> You can migrate a small set of users on the development instance for testing/staging. To enable importing users to your development instance, add `IMPORT_TO_DEV_INSTANCE=true` to the `.env` for the migration script.

Reviewer: We capitalize and add a code snippet for `Development` and `Production` in [this guide](https://clerk.com/docs/deployments/environments#instances-environments), so should we consistent with that? What do you think?

```suggestion
Every Clerk application has a `Development` and a `Production` instance. Before you start migrating user data, you need to configure your Clerk `Production` instance and migrate your Auth.js users directly into that instance. The [Deploying to Production](/docs/deployments/overview#deploying-to-production) page covers creating a `Production` instance.

You can migrate a small set of users on the `Development` instance for testing/staging. To enable importing users to your `Development` instance, add `IMPORT_TO_DEV_INSTANCE=true` to the `.env` for the migration script.
```

> ### Migrate User Data

Reviewer: ```suggestion
### Migrate user data from Auth.js to Clerk
```

This walkthrough will help you move user data from your existing database to Clerk.

To retain the user data in your database for easy querying, check out the documentation on [data synchronization with webhooks](/docs/users/sync-data).

1. Clone `github.com/clerk/migration-script`

> 2. Create a `.env` file in the root of the cloned repository with the `CLERK_SECRET_KEY` **of your production instance**.

Reviewer: ```suggestion
2. Create a `.env` file in the root of the cloned repository with the `CLERK_SECRET_KEY` **of your `Production` instance**.
```

> 3. Export all the user data from your database into a users.json file. The file should be in the following format.

Reviewer: ```suggestion
3. Export all the user data from your database into a `users.json` file. The file should be in the following format:
```

```json filename="users.json"
[
  {
    "userId": "string",
    "email": "email",
    "firstName": "string (optional)",
    "lastName": "string (optional)",
    "password": "string (optional)",
    "passwordHasher": "argon2 | argon | bcrypt | md5 | pbkdf2_sha256 | pbkdf2_sha256_django | pbkdf2_sha1 | scrypt_firebase",
  }
]
```

> 4. If you already have an API endpoint in your Auth.js app that returns a list of users you could use that. Otherwise you will need to query your database to obtain the user information, or use an export function from a database management tool.

Reviewer: ```suggestion
4. If you already have an API endpoint in your Auth.js app that returns a list of users, you can use that. Otherwise, you will need to query your database to obtain the user information, or use an export function from a database management tool.

The example below is a SQL query that would return the user information in the correct format for the migration script.
```

    
```sql
  SELECT id as userId, email, name FROM users
```
    
5. Edit the `.env.` file in the migration, and add your Clerk Secret Key using `CLERK_SECRET_KEY`
6. Run the script with `npm start`

> 7. Verify by signing in to your application secured by Clerk, and by checking the users listed in the [Users](https://dashboard.clerk.com/last-active?path=users) section of the Clerk Dashboard.

Reviewer: I think this is what you mean? The user should sign in to their instance (creating their user) and then they can check the dashboard that their user successfully exists as a user of the instance??

```suggestion
7. Verify the migration to Clerk was successful. Sign in to your application secured by Clerk. Then, check that your user is listed in the [Users](https://dashboard.clerk.com/last-active?path=users) section of the Clerk Dashboard.
```

Reviewer 2: It doesn't need to be in either order. I will probably change it to check the Dashboard first as that makes more sense.

9. Check for an error log for any users that were not migrated successfully.

> ### Further Support

Reviewer: ```suggestion
### Finding further support for migrating from Auth.js to Clerk
```

> This guide covers the most common steps that you would take for the migration. If you have more complex integrations with Auth that are not covered here, don't hesitate to hop into the Clerk discord and open a support question.

Reviewer: ```suggestion
This guide covers the most common steps that you would take for the migration. If you have more complex integrations with Auth.js that are not covered here, don't hesitate to reach out in the [Clerk Discord](https://discord.com/invite/b5rXHjAg7A) by creating a post in the [Support channel](https://discord.gg/bmPVkeqKzZ).
```

// See https://clerk.com/docs/references/nextjs/auth-middleware for more information about configuring your Middleware
export default authMiddleware({});
 
export const config = {
  matcher: ["/((?!.+\\.[\\w]+$|_next).*)", "/", "/(api|trpc)(.*)"],
};
```

#### Control access to your app with Clerk's components

To conditionally render UI when the user is signed in, wrap it with [`<SignedIn>`](/docs/components/control/signed-in).

To conditionally render UI when the user is _not_ signed in, wrap it with [`<SignedOut>`](/docs/components/control/signed-out).

```tsx filename="app/page.tsx"
import { SignedIn, SignedOut } from "@clerk/nextjs";
 
export default function Home() {
  return (
    <div>
      <SignedOut>
        <p>This content is public. Only signed out users can see this.</p>
      </SignedOut>
      <SignedIn>
        <p>This content is private. Only signed in users can see this.</p>
      </SignedIn>
    </div>
  )
}
```

### Read user and session data

#### Server side

Replace any Auth.js `getServerSession(req, res, authOptions)` with Clerk's helpers.

Let's see what your code might look like with Auth.js's `getServerSession()`:

> ```tsx filename="app/page.tsx"

Reviewer: We have App Router & Pages Router examples all thru out this guide. We should stay consistent and show here what the `getServerSession()` usage would like in their Pages Router app

Reviewer 2: This is leftover and needs deleting. The examples for both are below in Tabs.

Reviewer 2: This is the Auth.js code. Based on your previous edits I think we want to delete this.

Reviewer: omg you know what, I just thought of something - lennart uses a `diff` syntax on the code blocks to create diffs ([example](https://github.com/clerk/clerk-docs/pull/605/files#diff-3e9e3463657da8ff372ec5ca21d680029e440e06db876ea3b9754225316eed5eR177)).

we could use that here, so show auth.js code you would remove , and clerk code that you would add.
what do you think??

Reviewer 2: We could. My feeling is that your first instinct to not include the Auth.js code is correct. Chances are good that the implementation is much more complex than the sample, so I think its limited value.

Reviewer: okay let's remove it!

export default function Page() {
  const data = await getServerSession(authOptions);
  console.log(data)

  return (
    <p>Home Page</p>
  )
}
```

Now let's see what your code should look like when replacing `getServerSession` with Clerk's helpers:

<Tabs type="installer" items={["App Router", "Pages Router"]}>
  <Tab>
  You can replace Auth.js's `setServerSession` with Clerk's [`auth()`](/docs/references/nextjs/auth#auth) helper in order to read your user data.

  ```tsx failename="app/page.tsx"
  import { auth, currentUser } from "@clerk/nextjs";
  
  export default async function Page() {
    const { userId } = auth();
    console.log(userId)
    
    return (
      <p>Home Page</p>
    )
  }
  ```
  </Tab>

  <Tab>
  You can replace Auth.js's `setServerSession` with Clerk's [`getAuth()`](/docs/references/nextjs/get-auth#get-auth) helper in order to read your user data.

  ```tsx filename="pages/index.tsx"
  export async function getServerSideProps(context) {
    const session = getAuth(context.req);
    
    return { props: { ...buildClerkProps(ctx.req) } };
  };
  ```
  </Tab>
</Tabs>

#### Client Side

Replace Auth.js's `useSession()` hook with Clerk's hooks.

> Let's see what your code might look like with Auth.js's `useSession()`:

Reviewer: We have App Router & Pages Router examples all thru out this guide. We should stay consistent and show here what the `useSession()` usage would like in their Pages Router app

Reviewer 2: Its the basically the same. Remove `'use client'` and change the filename. I didn't include as we are working to de-dupe code.

Reviewer: ohh you know what you're right

#### Control access to your app with Clerk Middleware

You will need to remove the Auth.js Middleware from your application, and replace it with the Clerk's Middleware helper, `authMiddleware()`. 

Auth.js's middleware always rejects unauthorized requests. You may have additionally configured the Next.js middleware config to protect specific private routes. You will need to make note of this configuration so you can recreate it.

Clerk's Middleware gives you fine-grained control over handling the authenticated state and will, by default, run for your entire application.

> The example below is a basic configuration that requires authentication for all routes. You will need to edit this to allow specific routes to be public, as needed. Read the [`authMiddleware()`](/docs/references/nextjs/auth-middleware) documentation to learn more about how you can configure your Middleware.

Reviewer 2: "requires authentication for" isn't strictly true for Auth.js. It can work differently.

Reviewer: but its for the authMiddleware example, not Auth.js

Reviewer 2: My bad. Too much time staring at this guide. Change applied.
